R/carehomes_rt.R

```diff
@@ -347,69 +463,52 @@ carehomes_Rt_mean_duration_weighted_by_infectivity <- function(step, pars,
-
-  mean_duration_I_C_2 <- array(pars$I_C_2_transmission * p_C *
-    pars$k_C_2, dims) / denom(pars$gamma_C_2)
-
-  mean_duration_G_D <- array(pars$G_D_transmission * p_C * p_H *
-    p_G_D * pars$k_G_D / (1 - exp(-dt * pars$gamma_G_D)), dims)
-
-  mean_duration_hosp <- array(pars$hosp_transmission * (
-    prob_H_R * pars$k_H_R / (1 - exp(-dt * pars$gamma_H_R)) +
-      prob_H_D * pars$k_H_D / (1 - exp(-dt * pars$gamma_H_D)) +
-      (prob_ICU_W_R + prob_ICU_W_D + prob_ICU_D) * pars$k_ICU_pre /
-        (1 - exp(-dt * pars$gamma_ICU_pre))), dims)
+  calculate_mean <- function(par, prob, gamma) {
+    aperm(aperm(par * prob, c(4, 1, 2, 3)) / pexp(gamma, dt), c(2, 3, 4, 1))
+  }
 
-  mean_duration_icu <- array(pars$ICU_transmission * (
-    prob_ICU_W_R * pars$k_ICU_W_R / (1 - exp(-dt * pars$gamma_ICU_W_R)) +
-      prob_ICU_W_D * pars$k_ICU_W_D / (1 - exp(-dt * pars$gamma_ICU_W_D)) +
-      prob_ICU_D * pars$k_ICU_D / (1 - exp(-dt * pars$gamma_ICU_D))), dims)
-  ## safety check
-  stopifnot(identical(mean_duration_icu[, , , 1],
-                      mean_duration_icu[, , , n_strains]))
+  mean_duration_I_A <- calculate_mean(pars$I_A_transmission,
+                                      (1 - p_C) * pars$k_A, pars$gamma_A)
+  mean_duration_I_P <- calculate_mean(pars$I_P_transmission,
+                                      p_C * pars$k_P, pars$gamma_P)
+  mean_duration_I_C_1 <- calculate_mean(pars$I_C_1_transmission,
+                                        p_C * pars$k_C_1, pars$gamma_C_1)
+  mean_duration_I_C_2 <- calculate_mean(pars$I_C_2_transmission,
+                                        p_C * pars$k_C_2, pars$gamma_C_2)
+  mean_duration_G_D <- calculate_mean(pars$G_D_transmission,
+                                      p_C * p_H * p_G_D * pars$k_G_D,
+                                      pars$gamma_G_D)
+
+  mean_duration_hosp <-
+    pars$hosp_transmission *
+    (calculate_mean(pars$k_H_R, prob_H_R, pars$gamma_H_R) +
+       calculate_mean(pars$k_H_D, prob_H_D, pars$gamma_H_D) +
+       calculate_mean(pars$k_ICU_pre, prob_ICU_W_R + prob_ICU_W_D + prob_ICU_D,
+                      pars$gamma_ICU_pre))
+
+  mean_duration_icu <-
+    pars$ICU_transmission *
+    (calculate_mean(pars$k_ICU_W_R, prob_ICU_W_R, pars$gamma_ICU_W_R) +
+       calculate_mean(pars$k_ICU_W_D, prob_ICU_W_D, pars$gamma_ICU_W_D) +
+       calculate_mean(pars$k_ICU_D, prob_ICU_D, pars$gamma_ICU_D))
 
   mean_duration <- mean_duration_I_A + mean_duration_I_P +
     mean_duration_I_C_1 + mean_duration_I_C_2 + mean_duration_G_D +
     mean_duration_hosp + mean_duration_icu
 
   ## Account for different infectivity levels depending on vaccination stage
-
   mean_duration <- mean_duration *
     outer(outer(pars$rel_infectivity, rep(1, n_time_steps)), rep(1, n_strains))
 
+  ## Account for different infectivity levels depending on strain
+  mean_duration <-
+    aperm(aperm(mean_duration, c(4, 1, 2, 3)) * pars$strain_transmission,
+          c(2, 3, 4, 1))
+
```

Comment:
Not sure I understand what is going on with the `aperm`s here, so maybe useful to have a bit more commenting?

Reply:
Fixed in https://github.com/mrc-ide/sircovid/pull/272